### PR TITLE
Enhance unit testing for cred serialization

### PR DESCRIPTION
**Pull request by jpeeler**
Similar to the fuzz testing done in serializer_test, this tests the
serialize function at a higher level from the Inject prospective. The
other difference is that serialization is compared in one direction
only, so no unmarshalling is required to verify the result. Ints,
floats, strings, slices, and maps are data types tested to be correct so
far.

### CODE REVIEW

File: pkg/controller/injector/k8s_binding_injector.go [hunk 1]
```diff
@@ -54,7 +54,7 @@ func CreateK8sBindingInjector() (BindingInjector, error) {
 	}, nil
 }
 
-func (b *k8sBindingInjector) Inject(binding *servicecatalog.Binding, cred *brokerapi.Credential) error {
+func createSerializedSecret(binding *servicecatalog.Binding, cred *brokerapi.Credential) (*v1.Secret, error) {
```
Comment (MHBauer) [inline pkg/controller/injector/k8s_binding_injector.go]: refactor to allow direct access to the function, correct?

File: pkg/controller/injector/k8s_binding_injector_test.go [hunk 2]
```diff
@@ -29,6 +30,123 @@ import (
...
+				},
+			},
+		},
+		{
+			name: "map type",
+			cred: &brokerapi.Credential{
+				"Hostname": "host",
+				"Port":     "123",
+				"Username": "user",
+				"Password": map[string]int{"one": 1, "two": 2},
```
Comment (MHBauer) [inline pkg/controller/injector/k8s_binding_injector_test.go]: What is the purpose of this case? Where would we have maps? Is there a better field that would actually be a map?

File: pkg/controller/injector/k8s_binding_injector_test.go [hunk 3]
```diff
@@ -29,6 +30,123 @@ import (
...
+				"Port":     "123",
+				"Username": "user",
+				"Password": 1.23,
+			},
+			secret: &v1.Secret{
+				Data: map[string][]byte{
+					"Hostname": []byte("host"),
+					"Port":     []byte("123"),
+					"Username": []byte("user"),
+					"Password": []byte("1.23"),
```
Comment (MHBauer) [inline pkg/controller/injector/k8s_binding_injector_test.go]: What is the purpose of this test case?

Do we need to worry about numbers with exponential notation?

File: pkg/controller/injector/k8s_binding_injector_test.go [hunk 4]
```diff
@@ -29,6 +30,123 @@ import (
...
+				},
+			},
+		},
+		{
+			name: "int type",
+			cred: &brokerapi.Credential{
+				"Hostname": "host",
+				"Port":     123,
```
Comment (MHBauer) [inline pkg/controller/injector/k8s_binding_injector_test.go]: I know ports only go up to 2^16, but do we need to worry about giant numbers that become exponents in javascript?

**Comment by krancour** [comment]
> I am not sure I fully understand why we're testing what is being tested here.

Let me  say that I don't understand it either. I _do_ know that @pmorie wanted this, but it still doesn't make sense to me.

> this tests the serialize function at a higher level

What's the point of that though? These are more complicated tests to assert something that existing, more focused tests have already told us-- that serialization works.

File: pkg/controller/injector/k8s_binding_injector.go (hunk 1 above)
Comment (jpeeler) [inline pkg/controller/injector/k8s_binding_injector.go]: Yes, specifically to avoid any client operations with the secret.

File: pkg/controller/injector/k8s_binding_injector_test.go (hunk 3 above)
Comment (jpeeler) [inline pkg/controller/injector/k8s_binding_injector_test.go]: The password field is ensuring floats are serialized properly (the name in each loop is describing the test purpose). I don't know if exponential notation should be supported or not, but it doesn't hurt to verify.

File: pkg/controller/injector/k8s_binding_injector_test.go (hunk 4 above)
Comment (jpeeler) [inline pkg/controller/injector/k8s_binding_injector_test.go]: Verifying proper handling of overflow seems reasonable.

File: pkg/controller/injector/k8s_binding_injector_test.go (hunk 2 above)
Comment (jpeeler) [inline pkg/controller/injector/k8s_binding_injector_test.go]: The purpose of all the cases are to ensure serialization works with different types, not based on the expected values of data in today's implementation.

File: pkg/controller/injector/k8s_binding_injector_test.go (hunk 3 above)
Comment (krancour) [inline pkg/controller/injector/k8s_binding_injector_test.go]: Verifying transformations at such a low level seems it's more appropriately in the domain of the serialization tests. That said, the fuzzers used therein should already have us covered on that front.

File: pkg/controller/injector/k8s_binding_injector_test.go (hunk 4 above)
Comment (krancour) [inline pkg/controller/injector/k8s_binding_injector_test.go]: > but do we need to worry about giant numbers that become exponents in javascript?

The fact that this is a port is actually inconsequential. It's example usage of an integer. The secrets that get injected could very well contain integers that are something other than a port.

That being  said, the serialization tests already cover this.

File: pkg/controller/injector/k8s_binding_injector_test.go (hunk 2 above)
Comment (krancour) [inline pkg/controller/injector/k8s_binding_injector_test.go]: This cannot actually occur in our system. Credentials are unmarshalled from JSON in the binding response from an OSB broker. Maps get unmarshalled to `map[string]interface{}`, so a credential with `map[string]int` simply cannot occur.

If this _were_ a valid thing to test, this scenario should be added to the serialization tests; not here.

File: pkg/controller/injector/k8s_binding_injector_test.go (hunk 3 above)
Comment (pmorie) [inline pkg/controller/injector/k8s_binding_injector_test.go]: There is business logic in this code that needs to be tested.  Right now the business logic is simple - it's 'put everything in this deserialized response into a secret'.  Business logic is going to accumulate in this code, however - we're going to eventually be putting different pieces of information into different resources.  That is why we should test this code.

**Comment by pmorie** [comment]
> this tests the serialize function at a higher level

I would instead characterize this as tests for the business logic of how a deserialized binding response is transformed into resources.  It is very simple business logic now, but it will grow to be more complex.

File: pkg/controller/injector/k8s_binding_injector_test.go (hunk 3 above)
Comment (krancour) [inline pkg/controller/injector/k8s_binding_injector_test.go]: > That is why we should test this code.

I'm not at all saying we shouldn't test this code. I'm only arguing for the _right_ tests. Before this PR, the tests in `k8s_binding_injector_test.go` were at the right level. They tested that when injection occurs, the resulting secret has in it all the same data as the credentials object. i.e. It tested that _injection_ works and it does so agnostic of what underlying serialization process is in play. The more complex bit of ensuring that serialization process works was left to its own tests in `serializer_test.go`.

After this PR, we are testing serialization _twice_ and adding no further value because the "other business logic" you say will one day exist in the injector currently doesn't. When the day comes, _yes_, we do need to amend these tests to test that new business logic, but even then, I can't see what benefit arises from duplicate serialization tests. When more business log gets added, the new tests that are added along with it shouldn't focus on serialization, but should focus on the business logic.

**Comment by MHBauer** [comment]
Is there a particular reason we only touch the password field? From what I can tell it's not treated specially. Shove the same input into every field.

I get that there might be logic in here in the future, but right now we're mainly just testing serialize, which we already have tests for.

I would expect one test case, to make sure the output object is formatted correctly.

If we're going to "test serialzation" again, then set all the fields in each test case.

It's not clear from reading the test what's going on. More comments or more descriptive 'name' might help. Or change the name field to have a more descriptive name.

File: pkg/controller/injector/k8s_binding_injector.go [hunk 5]
```diff
@@ -67,12 +67,22 @@ func (b *k8sBindingInjector) Inject(binding *servicecatalog.Binding, cred *broke
 		var err error
 		secret.Data[k], err = serialize(v)
 		if err != nil {
-			return fmt.Errorf("Unable to serialize credential value %q: %v; %s",
+			return nil, fmt.Errorf("Unable to serialize credential value %q: %v; %s",
```
Comment (arschles) [inline pkg/controller/injector/k8s_binding_injector.go]: it would be nice to split this `fmt.Errorf` into an error type

File: pkg/controller/injector/k8s_binding_injector_test.go (hunk 2 above)
Comment (duglin) [inline pkg/controller/injector/k8s_binding_injector_test.go]: I think the fact its an "int" instead of an "interface{}" doesn't really matter much. To me the key aspect is that map is being used at all.

**Comment by duglin** [comment]
even if there's overlap with some other tests, I'm ok with this because its testing it at another level in the system - which helps make sure nothing is tweakings things along the way. Or, if I'm wrong, we just have more tests - which is rarely a bad thing :-) 

Merging.  LGTM
